particles/Tutorial/Javascript/Demo/source/MoveApplier.js

```diff
@@ -0,0 +1,104 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+ /* global defineParticle */
+
+ defineParticle(({DomParticle, html}) => {
+
+  return class extends DomParticle {
+    get template() {
+      return '';
+    }
+
+    update({move, gameState}, {}) {
+      if (!gameState) { // Gamestate has not been set yet
+        const gs = {
```

Comment:
I am uncomfortable about having so much derived state and not having an intuitive understand of what some of these attributes are.

Derived state is where you store state that is derived from another state. It is dangerous because 2 pieces of state can get out of sync and it's a great opportunity for making bugs. Having reactive-style function between the 2 pieces of state fixes the problem, but we don't do that here, we just manage things separately.

E.g. `gameOver` and `winnerAvatar` is derived from `board` (maybe `moves` as well, depending on what is stored there, I'm not sure, certainly you can read the number of moves made from the board), ergo it doesn't need to be stored. That's a bit easier said than done, as (I think) we don't have easy way to share code between JS particles (to be checked with Scott), if we did I would recommend having a helper function. In Kotlin it actually should be easier, there you just compile a Kotlin file into multiple wasm modules.

But maybe... just maybe, you only care about computing the gameOver in one place, and the code that does that can live there - to be investigated.

Now regarding intuitive understanding and good naming: What is `moves` vs. `attemptedMoves` ? I would need to read the code _very_ carefully to understand that, which is suboptimal.

We have at least 2 options: You can try to simplify it yourselves, or I can try to simplify it with you. We may find that it is as simple as it can possibly be, but I doubt that. Writing code is like... writing words. Making this small and to the point takes time. And a thing once written is read many times, especially a tutorial, so I'm sure this effort would pay off.

Reply:
I've gotten rid of moves and attemptedMoves entirely. I think there's a lot less derived state with the new refactor, thoughts?